When using InterceptionProxy to create multiple beans of the same type `IllegalStateException: Multiple GeneratedClassBuildItem were produced for the same classes`

### Describe the bug

To get CDI Interceptors working with beans created with `@Produces` (based on this comment https://github.com/quarkusio/quarkus/issues/44815#issuecomment-2507302091) we wrapped those beans with an `InterceptionProxy`, which works fine in quarkus:dev mode.

When fully building an uber jar though, we get an error like this for any proxied bean that has multiple versions.

E.g. a sample class:
```
@ApplicationScoped
public class Config {
    @Produces
    @ApplicationScoped
    @Named("version1")
    public ExampleClass version1(InterceptionProxy<ExampleClass> proxy) {
        return proxy.create(new ExampleClass());
    }
    
    @Produces
    @ApplicationScoped
    @Named("version2")
    public ExampleClass version2(InterceptionProxy<ExampleClass> proxy) {
        return proxy.create(new ExampleClass());
    }
}
```


```
[ERROR] 	[error]: Build step io.quarkus.deployment.pkg.steps.JarResultBuildStep#buildRunnerJar threw an exception: IllegalStateException: Multiple GeneratedClassBuildItem were produced for the same classes:
[ERROR]
[ERROR] - com.example.package.ExampleClass_InterceptionSubclass: 2
[ERROR]
[ERROR] 	at io.quarkus.deployment.pkg.jar.AbstractJarBuilder.checkConsistency(AbstractJarBuilder.java:92)
[ERROR] 	at io.quarkus.deployment.pkg.jar.AbstractJarBuilder.<init>(AbstractJarBuilder.java:76)
[ERROR] 	at io.quarkus.deployment.pkg.jar.UberJarBuilder.<init>(UberJarBuilder.java:77)
[ERROR] 	at io.quarkus.deployment.pkg.steps.JarResultBuildStep.buildRunnerJar(JarResultBuildStep.java:120)
[ERROR] 	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:872)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:255)
[ERROR] 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1622)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1589)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:1583)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
```

Seems like this is probably a tricky problem to solve (proxies need to include their beans name in the class or something?), but it is the main use case I have personally for using the `@Produces` pattern to create multiple flavors of the same bean with different config or when using third party classes.

For now I'm just adding wrapper classes with different types to work around it.

### Expected behavior

Be able to create multiple versions of the same class with `InterceptionProxy`

### Actual behavior

buildRunnerJar throws an error during clean install.

### How to Reproduce?

Use InterceptorProxy to create 2 beans with the same class. Works in quarkus dev mode, fails when running an uber jar.

I can create a sample project if desired, didn't want to waste time if this is expected behavior with no plan to fix.

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/InterceptionProxyGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/InterceptionProxyGenerator.java
@@ -82,7 +82,7 @@ public class InterceptionProxyGenerator extends AbstractGenerator {
 
         createInterceptionProxyProvider(gizmo, bean);
         createInterceptionProxy(gizmo, bean);
-        createInterceptionSubclass(gizmo, bean.getInterceptionProxy(), bytecodeTransformerConsumer,
+        createInterceptionSubclass(gizmo, bean, bytecodeTransformerConsumer,
                 transformUnproxyableClasses);
 
         return classOutput.getResources();
@@ -98,8 +98,8 @@ public class InterceptionProxyGenerator extends AbstractGenerator {
         return bean.getBeanClass().toString() + "_InterceptionProxy_" + bean.getIdentifier();
     }
 
-    private static String interceptionSubclassName(InterceptionProxyInfo interceptionProxy) {
-        return interceptionProxy.getTargetClass() + INTERCEPTION_SUBCLASS;
+    private static String interceptionSubclassName(InterceptionProxyInfo interceptionProxy, BeanInfo bean) {
+        return interceptionProxy.getTargetClass() + "_" + bean.getIdentifier() + INTERCEPTION_SUBCLASS;
     }
 
     private void createInterceptionProxyProvider(Gizmo gizmo, BeanInfo bean) {
@@ -157,7 +157,8 @@ public class InterceptionProxyGenerator extends AbstractGenerator {
                 mc.body(b0 -> {
                     InterceptionProxyInfo interceptionProxy = bean.getInterceptionProxy();
                     b0.ifInstanceOf(delegate, classDescOf(interceptionProxy.getTargetClass()), (b1, ignored) -> {
-                        ConstructorDesc ctor = ConstructorDesc.of(ClassDesc.of(interceptionSubclassName(interceptionProxy)),
+                        ConstructorDesc ctor = ConstructorDesc.of(
+                                ClassDesc.of(interceptionSubclassName(interceptionProxy, bean)),
                                 CreationalContext.class, Object.class);
                         b1.return_(b1.new_(ctor, cc.this_().field(ccField), delegate));
                     });
@@ -174,8 +175,9 @@ public class InterceptionProxyGenerator extends AbstractGenerator {
         });
     }
 
-    private void createInterceptionSubclass(Gizmo gizmo, InterceptionProxyInfo interceptionProxy,
+    private void createInterceptionSubclass(Gizmo gizmo, BeanInfo bean,
             Consumer<BytecodeTransformer> bytecodeTransformerConsumer, boolean transformUnproxyableClasses) {
+        InterceptionProxyInfo interceptionProxy = bean.getInterceptionProxy();
         BeanInfo pseudoBean = interceptionProxy.getPseudoBean();
         ClassDesc pseudoBeanClass = classDescOf(pseudoBean.getImplClazz());
         boolean implementingInterface = pseudoBean.getImplClazz().isInterface();
@@ -184,7 +186,7 @@ public class InterceptionProxyGenerator extends AbstractGenerator {
 
         ClassDesc superClass = implementingInterface ? CD_Object : pseudoBeanClass;
 
-        gizmo.class_(interceptionSubclassName(interceptionProxy), cc -> {
+        gizmo.class_(interceptionSubclassName(interceptionProxy, bean), cc -> {
             cc.extends_(superClass);
             if (implementingInterface) {
                 cc.implements_(pseudoBeanClass);

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/producer/InterceptionProxyMultipleBeansSameTypeTest.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/producer/InterceptionProxyMultipleBeansSameTypeTest.java
@@ -1,0 +1,143 @@
+package io.quarkus.arc.test.interceptors.producer;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertInstanceOf;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNotSame;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import jakarta.annotation.Priority;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.enterprise.context.Dependent;
+import jakarta.enterprise.inject.Produces;
+import jakarta.enterprise.util.AnnotationLiteral;
+import jakarta.inject.Qualifier;
+import jakarta.interceptor.AroundInvoke;
+import jakarta.interceptor.Interceptor;
+import jakarta.interceptor.InterceptorBinding;
+import jakarta.interceptor.InvocationContext;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.ArcContainer;
+import io.quarkus.arc.BindingsSource;
+import io.quarkus.arc.ClientProxy;
+import io.quarkus.arc.InterceptionProxy;
+import io.quarkus.arc.InterceptionProxySubclass;
+import io.quarkus.arc.test.ArcTestContainer;
+
+public class InterceptionProxyMultipleBeansSameTypeTest {
+    @RegisterExtension
+    public ArcTestContainer container = new ArcTestContainer(MyBinding.class, MyInterceptor.class, MyQualifier.class,
+            MyProducer.class, MyNonBeanBindings.class);
+
+    @Test
+    public void test() {
+        ArcContainer container = Arc.container();
+        MyNonbean nonbeanA = container.select(MyNonbean.class, new MyQualifier.Literal("A")).get();
+        assertNonbean(nonbeanA, "intercepted: A");
+        MyNonbean nonbeanB = container.select(MyNonbean.class, new MyQualifier.Literal("B")).get();
+        assertNonbean(nonbeanB, "B");
+    }
+
+    private void assertNonbean(MyNonbean nonbean, String expected) {
+        assertEquals(expected, nonbean.hello());
+
+        assertInstanceOf(ClientProxy.class, nonbean);
+        MyNonbean unwrappedProxy = ClientProxy.unwrap(nonbean);
+        assertNotNull(unwrappedProxy);
+        assertNotSame(nonbean, unwrappedProxy);
+
+        assertInstanceOf(InterceptionProxySubclass.class, unwrappedProxy);
+        assertNotNull(InterceptionProxySubclass.unwrap(unwrappedProxy));
+        assertNotSame(unwrappedProxy, InterceptionProxySubclass.unwrap(unwrappedProxy));
+        assertNotSame(nonbean, InterceptionProxySubclass.unwrap(unwrappedProxy));
+    }
+
+    @Retention(RetentionPolicy.RUNTIME)
+    @Qualifier
+    @interface MyQualifier {
+
+        String value();
+
+        public final static class Literal extends AnnotationLiteral<MyQualifier> implements MyQualifier {
+
+            private final String value;
+
+            Literal(String value) {
+                this.value = value;
+            }
+
+            @Override
+            public String value() {
+                return value;
+            }
+
+        }
+
+    }
+
+    @Retention(RetentionPolicy.RUNTIME)
+    @Target({ ElementType.TYPE, ElementType.METHOD, ElementType.CONSTRUCTOR })
+    @InterceptorBinding
+    @interface MyBinding {
+    }
+
+    @MyBinding
+    @Priority(1)
+    @Interceptor
+    static class MyInterceptor {
+        @AroundInvoke
+        Object intercept(InvocationContext ctx) throws Exception {
+            return "intercepted: " + ctx.proceed();
+        }
+    }
+
+    static class MyNonbean {
+
+        private final String msg;
+
+        MyNonbean() {
+            this.msg = null;
+        }
+
+        MyNonbean(String msg) {
+            this.msg = msg;
+        }
+
+        String hello() {
+            return msg;
+        }
+    }
+
+    static abstract class MyNonBeanBindings {
+
+        @MyBinding
+        abstract String hello();
+
+    }
+
+    @Dependent
+    static class MyProducer {
+
+        @Produces
+        @MyQualifier("A")
+        @ApplicationScoped
+        MyNonbean produceA(@BindingsSource(MyNonBeanBindings.class) InterceptionProxy<MyNonbean> proxy) {
+            return proxy.create(new MyNonbean("A"));
+        }
+
+        @Produces
+        @MyQualifier("B")
+        @ApplicationScoped
+        MyNonbean produceB(InterceptionProxy<MyNonbean> proxy) {
+            return proxy.create(new MyNonbean("B"));
+        }
+    }
+}

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/producer/ProducerWithInterceptionAndVirtualCallInCtorTest.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/producer/ProducerWithInterceptionAndVirtualCallInCtorTest.java
@@ -37,7 +37,7 @@ public class ProducerWithInterceptionAndVirtualCallInCtorTest {
 
         assertEquals(List.of(
                 "MyNonbean",
-                "ProducerWithInterceptionAndVirtualCallInCtorTest$MyNonbean_InterceptionSubclass"),
+                nonbean.getClass().getSimpleName()),
                 MyNonbean.constructedInstances_intercepted);
         assertEquals(MyNonbean.constructedInstances_intercepted, MyNonbean.constructedInstances_notIntercepted);
     }

